boxsdk/object/upload_session.py

```diff
@@ -0,0 +1,165 @@
+            `list` of `dict`
+        """
+        return LimitOffsetBasedDictCollection(
+            session=self.session,
+            url=self.get_url('parts'),
+            limit=limit,
+            offset=offset,
+            fields=None,
+            return_full_pages=False,
+        )
+
+    def upload_part_bytes(self, part_bytes, offset, total_size, part_content_sha1=None):
+        """
+        Upload a part of a file.
+
+        :param part_bytes:
+            Part bytes
+        :type part_bytes:
+            `bytes`
+        :param offset:
+            Offset, in number of bytes, of the part compared to the beginning of the file. This number should be a
+            multiple of the part size.
+        :type offset:
+            `int`
+        :param total_size:
+            The size of the file that this part belongs to.
+        :type total_size:
+            `int`
+        :param part_content_sha1:
+            SHA-1 hash of the part's content. If not specified, this will be calculated.
+        :type part_content_sha1:
+            `unicode` or None
+        :returns:
+            The uploaded part record.
+        :rtype:
+            `dict`
+        """
+
+        if part_content_sha1 is None:
+            sha1 = hashlib.sha1()
+            sha1.update(part_bytes)
+            part_content_sha1 = sha1.digest()
+
+        range_end = min(offset + self.part_size - 1, total_size - 1)  # pylint:disable=no-member
+        headers = {
+            'Content-Type': 'application/octet-stream',
+            'Digest': 'SHA={0}'.format(base64.b64encode(part_content_sha1).decode('utf-8')),
+            'Content-Range': 'bytes {0}-{1}/{2}'.format(offset, range_end, total_size),
+        }
+        response = self._session.put(
+            self.get_url(),
+            headers=headers,
+            data=part_bytes,
+        )
+        return response.json()['part']
+
+    def commit(self, content_sha1, parts=None, file_attributes=None, etag=None):
+        """
+        Commit a multiput upload.
+
+        :param content_sha1:
+            SHA-1 hash of the file contents that was uploaded.
+        :type content_sha1:
+            `bytes`
+        :param parts:
+            List of parts that were uploaded.
+        :type parts:
+            `Iterable` of `dict` or None
+        :param file_attributes:
+            A `dict` of attributes to set on the uploaded file.
+        :type file_attributes:
+            `dict`
+        :param etag:
+            If specified, instruct the Box API to delete the folder only if the current version's etag matches.
+        :type etag:
+            `unicode` or None
+        :returns:
+            The newly-uploaded file object.
+        :rtype:
+            :class:`File`
+        """
+        body = {}
+        parts_list = []
+        if file_attributes is not None:
+            body['attributes'] = file_attributes
+        if parts is not None:
+            body['parts'] = parts
+        else:
```

Comment:
In your manual test we got things working by keeping track of the parts (and so using the `if` branch above).  Have you tested to verify that the `else` branch here works correctly?

Reply:
Just manually tested the else and the get_parts() works as well for the commit